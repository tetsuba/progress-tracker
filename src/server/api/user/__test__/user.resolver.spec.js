const { createTestServer } = require('../../../test/serverTestHelper')
const {
  LOGIN_USER_MUTATION,
  REGISTER_NEW_USER_MUTATION,
  REQUEST_PASSWORD_RESET_MUTATION,
  VERIFY_USER_EMAIL_MUTATION,
  RESET_USER_PASSWORD_MUTATION,
  UPDATE_USER_DETAILS_MUTATION,
} = require('../../../../client/api/user/user.mutation')
const {
  GET_USER_DETAILS_QUERY,
  IS_USER_SESSION_EXPIRED,
} = require('../../../../client/api/user/user.query')
const {
  userMockSuccess,
  userMockSuccessNotVerified,
  userMockErrorUserNotFound,
} = require('../../../../test/mockApi/user/userMockData')

const {
  tokenMockSuccess,
} = require('../../../../test/mockApi/token/tokenMockData')

describe('user.resolver', () => {
  describe('@Query', () => {
    describe('getUserDetails', () => {
      const ctx = {
        user: { id: 'userId1234' },
        models: {
          User: {
            findById: jest.fn(() => userMockSuccess),
          },
        },
      }

      test('should return user details', async () => {
        const { query } = createTestServer(ctx)
        const res = await query({
          query: GET_USER_DETAILS_QUERY,
        })
        expect(res).toMatchSnapshot()
      })
    })
    describe('isUserSessionExpired', () => {
      test('should return success "token valid"', async () => {
        const ctx = {
          user: { id: 'userId1234' },
        }
        const { query } = createTestServer(ctx)
        const res = await query({
          query: IS_USER_SESSION_EXPIRED,
        })
        expect(res).toMatchSnapshot()
      })

      test('should return error token expired', async () => {
        const ctx = { user: {} }
        const { query } = createTestServer(ctx)
        const res = await query({
          query: IS_USER_SESSION_EXPIRED,
        })
        expect(res).toMatchSnapshot()
      })
    })
  })
  describe('@Mutation', () => {
    describe('registerNewUser', () => {
      const ctx = {
        models: {
          User: {
            findOne: jest.fn(() => userMockErrorUserNotFound),
            create: jest.fn(() => userMockSuccess),
            deleteOne: jest.fn(() => ({})),
          },
          Token: {
            create: jest.fn(() => tokenMockSuccess),
          },
        },
      }
      const mutations = {
        mutation: REGISTER_NEW_USER_MUTATION,
        variables: {
          input: {
            firstName: 'John',
            lastName: 'Doh',
            email: 'test1234@test.com',
            password: '1234qwer',
          },
        },
      }
      test('should register a new user returning a success message', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })

      test('should return an error if email is not valid address', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate({
          ...mutations,
          variables: {
            input: {
              ...mutations.variables.input,
              email: 'incorrect@format-com',
            },
          },
        })
        expect(res).toMatchSnapshot()
      })

      test('should return an error in email address already exist', async () => {
        const { mutate } = createTestServer({
          ...ctx,
          models: {
            ...ctx.models,
            User: {
              ...ctx.models.User,
              findOne: jest.fn(() => userMockSuccess),
            },
          },
        })
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })
    })
    describe('loginUser', () => {
      const ctx = {
        models: {
          User: {
            findOne: jest.fn(() => userMockSuccess),
          },
          Token: {
            create: jest.fn(() => tokenMockSuccess),
          },
        },
      }
      const mutations = {
        mutation: LOGIN_USER_MUTATION,
        variables: { input: { email: 'test@test.com', password: '1234qwer' } },
      }
      test('should return a token', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })

      test('should return an error if a user is not found', async () => {
        const { mutate } = createTestServer({
          models: {
            ...ctx.models,
            User: {
              findOne: jest.fn(() => userMockErrorUserNotFound),
            },
          },
        })
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })

      test('should return an error if passwords do not match', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate({
          ...mutations,
          variables: { input: { email: 'test@test.com', password: '1234' } },
        })
        expect(res).toMatchSnapshot()
      })

      test('should return an error if email is not verified', async () => {
        const { mutate } = createTestServer({
          models: {
            ...ctx.models,
            User: {
              findOne: jest.fn(() => userMockSuccessNotVerified),
            },
          },
        })
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })
    })
    describe('verifyUserEmail', () => {
      const ctx = {
        models: {
          User: {
            findOne: jest.fn(() => userMockSuccessNotVerified),
          },
          Token: {
            create: jest.fn(() => tokenMockSuccess),
          },
        },
      }
      const mutations = {
        mutation: VERIFY_USER_EMAIL_MUTATION,
        variables: { input: { email: 'test@test.com' } },
      }

      test('should return a success message', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })

      test('should return an error if a user email is not found', async () => {
        const { mutate } = createTestServer({
          ...ctx,
          models: {
            ...ctx.models,
            User: {
              findOne: jest.fn(() => userMockErrorUserNotFound),
            },
          },
        })
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })

      test('should return an error if the users email address is not a valid format', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate({
          ...mutations,
          variables: { input: { email: 'test@testcom' } },
        })
        expect(res).toMatchSnapshot()
      })

      test('should return an error if a user email has been verified already', async () => {
        const { mutate } = createTestServer({
          ...ctx,
          models: {
            ...ctx.models,
            User: {
              findOne: jest.fn(() => userMockSuccess),
            },
          },
        })
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })
    })
    describe('requestPasswordReset', () => {
      const ctx = {
        models: {
          User: {
            findOne: jest.fn(() => userMockSuccess),
          },
          Token: {
            create: jest.fn(() => tokenMockSuccess),
          },
        },
      }
      const mutations = {
        mutation: REQUEST_PASSWORD_RESET_MUTATION,
        variables: { input: { email: 'test@test.com' } },
      }

      test('should return a success message', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })

      test('should return an error if the users email address is not found ', async () => {
        const { mutate } = createTestServer({
          ...ctx,
          models: {
            ...ctx.models,
            User: {
              findOne: jest.fn(() => userMockErrorUserNotFound),
            },
          },
        })
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })

      test('should return an error if the users email address is not a valid format', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate({
          ...mutations,
          variables: { input: { email: 'test@testcom' } },
        })
        expect(res).toMatchSnapshot()
      })
    })
    describe('resetUserPassword', () => {
      const ctx = {
        models: {
          User: {
            findById: jest.fn(() => ({
              save: jest.fn(() => userMockSuccess),
            })),
          },
          Token: {
            findOne: jest.fn(() => tokenMockSuccess),
            deleteOne: jest.fn(() => 'success'),
          },
        },
      }
      const mutations = {
        mutation: RESET_USER_PASSWORD_MUTATION,
        variables: { input: { password: 'password', token: 'token' } },
      }

      test('should return a success message', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })
    })
    describe('updateUserDetails', () => {
      const ctx = {
        user: { id: '1234User' },
        models: {
          User: {
            findById: jest.fn(() => ({
              ...userMockSuccess,
              save: jest.fn(() => userMockSuccess),
            })),
          },
        },
      }
      const mutations = {
        mutation: UPDATE_USER_DETAILS_MUTATION,
        variables: { input: { firstName: 'John', lastName: 'Doh' } },
      }

      test('should return user details', async () => {
        const { mutate } = createTestServer(ctx)
        const res = await mutate(mutations)
        expect(res).toMatchSnapshot()
      })
    })
  })
})
