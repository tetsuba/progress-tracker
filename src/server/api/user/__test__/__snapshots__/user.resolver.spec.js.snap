// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user.resolver @Mutation loginUser should return a token 1`] = `
Object {
  "data": Object {
    "loginUser": Object {
      "token": "token",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation loginUser should return an error if a user is not found 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: INCORRECT_USER_DETAILS],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation loginUser should return an error if email is not verified 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: EMAIL_NOT_VERIFIED],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation loginUser should return an error if passwords do not match 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: INCORRECT_USER_DETAILS],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation registerNewUser should register a new user returning a success message 1`] = `
Object {
  "data": Object {
    "registerNewUser": Object {
      "success": "Registering a user in a test environment wil create and delete the account and will not send an confirmation email",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation registerNewUser should return an error if email is not valid address 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: NOT_VALID_EMAIL],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation registerNewUser should return an error in email address already exist 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: EMAIL_ALREADY_EXIST],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation requestPasswordReset should return a success message 1`] = `
Object {
  "data": Object {
    "requestPasswordReset": Object {
      "success": "Please check your email and click on the link to reset your password",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation requestPasswordReset should return an error if the users email address is not a valid format 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: NOT_VALID_EMAIL],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation requestPasswordReset should return an error if the users email address is not found  1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: EMAIL_DOES_NOT_EXIST],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation resetUserPassword should return a success message 1`] = `
Object {
  "data": Object {
    "resetUserPassword": Object {
      "success": "New password is saved",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation updateUserDetails should return an "UNAUTHORIZED" if user is not authenticated 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: UNAUTHORIZED],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation updateUserDetails should return user details 1`] = `
Object {
  "data": Object {
    "updateUserDetails": Object {
      "email": "test@test.com",
      "firstName": "John",
      "lastName": "Doh",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation verifyUserEmail should return a success message 1`] = `
Object {
  "data": Object {
    "verifyUserEmail": Object {
      "success": "Please check your email and confirm by pressing on the link.",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation verifyUserEmail should return an error if a user email has been verified already 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: EMAIL_VERIFIED],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation verifyUserEmail should return an error if a user email is not found 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: EMAIL_DOES_NOT_EXIST],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Mutation verifyUserEmail should return an error if the users email address is not a valid format 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: NOT_VALID_EMAIL],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Query getUserDetails should return an "UNAUTHORIZED" if user is not authenticated 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: UNAUTHORIZED],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Query getUserDetails should return user details 1`] = `
Object {
  "data": Object {
    "getUserDetails": Object {
      "email": "test@test.com",
      "firstName": "John",
      "lastName": "Doh",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Query isUserSessionExpired should return error token expired 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: TOKEN_EXPIRED],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`user.resolver @Query isUserSessionExpired should return success "token valid" 1`] = `
Object {
  "data": Object {
    "isUserSessionExpired": Object {
      "success": "valid",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
